### PR TITLE
8320069: RISC-V: Add Zcb instructions

### PR DESCRIPTION
**Pull request by robehn**
Hi, this is the instructions for zcb.

Due to over lack of infrastructure having multiple extension dependent instruction does not fit well.
Some of these compressed instructions are also missing 1 to 1 mapping, e.g. now we have a compressed not, but the corresponding instruction in uncompressed is still xor.
I think we need to do some rework here.

I also I don't like the macro expansion as it hopeless in debugger and 'IDE's (vim+rtags for me).
(macro stuff was originally done when templates where blacklisted in hotspot)

And I don't want an option for this, as zcb is coming in hwprobe, if you have compressed on you get them if they are supported (may depend on e.g. zbb).

I have done some modification since it passed tier1, so I'm running stuff over the weekend.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320069](https://bugs.openjdk.org/browse/JDK-8320069): RISC-V: Add Zcb instructions (**Enhancement** - P5)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Vladimir Kempik](https://openjdk.org/census#vkempik) (@VladimirKempik - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17122/head:pull/17122` \
`$ git checkout pull/17122`

Update a local copy of the PR: \
`$ git checkout pull/17122` \
`$ git pull https://git.openjdk.org/jdk.git pull/17122/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17122`

View PR using the GUI difftool: \
`$ git pr show -t 17122`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17122.diff">https://git.openjdk.org/jdk/pull/17122.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17122#issuecomment-1857934381)

### CODE REVIEW

**Comment by VladimirKempik** [comment]
We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ? https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880

**Comment by robehn** [comment]
> We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ?
> 
> https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880

No, you are correct I also think this is not optimal. I don't know the background, but it seems like this is the easiest way to add compressed transparently. But to fully utilize C instruction we should favor the x8->x15, so we often don't get C due to e.g. BCP is in x22. I think to be able to better utilize C we can't have it so transparent.

So here I just try to follow the current code, see how lw is changed to c_lw.

**Comment by VladimirKempik** [comment]
> > We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ?
> > https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880
> 
> No, you are correct I also think this is not optimal. I don't know the background, but it seems like this is the easiest way to add compressed transparently. But to fully utilize C instruction we should favor the x8->x15, we often don't get C due to e.g. BCP is in x22. I think to be able to better utilize C we can't have it so transparent.
> 
> So here I just try to follow the current code, see how lw is changed to c_lw.

Not exactly related to this PR, but I also saw a strange behaviour from MacroAssembler's lwu.
it was generating lw + and ( a kind of lwu emulation) instead of lwu

an example
```
  0.44%  ?  0x0000003fa46a86c8:   slli    t3,t3,0x20
   0.48%  ?  0x0000003fa46a86ca:   addi    t3,t3,-1
  ....
   3.11%  ?  0x0000003fa46a86dc:   lw    a0,0(t1)
   5.34%  ?  0x0000003fa46a86e0:   and    a0,a0,t3
```
Using Assembler::lwu directly resulted in a correctly generated lwu

**Comment by robehn** [comment]
> > > We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ?
> > > https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880
> > 
> > 
> > No, you are correct I also think this is not optimal. I don't know the background, but it seems like this is the easiest way to add compressed transparently. But to fully utilize C instruction we should favor the x8->x15, we often don't get C due to e.g. BCP is in x22. I think to be able to better utilize C we can't have it so transparent.
> > So here I just try to follow the current code, see how lw is changed to c_lw.
> 
> Not exactly related to this PR, but I also saw a strange behaviour from MacroAssembler's lwu. it was generating lw + and ( a kind of lwu emulation) instead of lwu
> 
> an example
> 
> ```
>   0.44%  ?  0x0000003fa46a86c8:   slli    t3,t3,0x20
>    0.48%  ?  0x0000003fa46a86ca:   addi    t3,t3,-1
>   ....
>    3.11%  ?  0x0000003fa46a86dc:   lw    a0,0(t1)
>    5.34%  ?  0x0000003fa46a86e0:   and    a0,a0,t3
> ```
> 
> Using Assembler::lwu directly resulted in a correctly generated lwu

Yes, I have seen similar things.
```
  0x00002aaabc9464fc:   addiw   ra,ra,-1365 # 0x00000000000aaaab
  0x00002aaabc946500:   slli    ra,ra,0xd
  0x00002aaabc946502:   addi    ra,ra,-929
  0x00002aaabc946506:   slli    ra,ra,0xd
  0x00002aaabc946508:   addi    ra,ra,456
  0x00002aaabc94650c:   jalr    ra
```
As jalr have as "111001000" would fit in the signed 12imm I think this maybe sub-optimal.

I can go over and fix them, I'll create jira.

**Comment by RealFYang** [comment]
> > > We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ?
> > > https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880
> > 
> > 
> > No, you are correct I also think this is not optimal. I don't know the background, but it seems like this is the easiest way to add compressed transparently. But to fully utilize C instruction we should favor the x8->x15, we often don't get C due to e.g. BCP is in x22. I think to be able to better utilize C we can't have it so transparent.
> > So here I just try to follow the current code, see how lw is changed to c_lw.
> 
> Not exactly related to this PR, but I also saw a strange behaviour from MacroAssembler's lwu. it was generating lw + and ( a kind of lwu emulation) instead of lwu
> 
> an example
> 
> ```
>   0.44%  ?  0x0000003fa46a86c8:   slli    t3,t3,0x20
>    0.48%  ?  0x0000003fa46a86ca:   addi    t3,t3,-1
>   ....
>    3.11%  ?  0x0000003fa46a86dc:   lw    a0,0(t1)
>    5.34%  ?  0x0000003fa46a86e0:   and    a0,a0,t3
> ```
> 
> Using Assembler::lwu directly resulted in a correctly generated lwu

Interesting. This does not seem to menifest on the code of `MacroAssembler's lwu`. I wonder how could that happens.

**Comment by VladimirKempik** [comment]
> > > > We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ?
> > > > https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880
> > > 
> > > 
> > > No, you are correct I also think this is not optimal. I don't know the background, but it seems like this is the easiest way to add compressed transparently. But to fully utilize C instruction we should favor the x8->x15, we often don't get C due to e.g. BCP is in x22. I think to be able to better utilize C we can't have it so transparent.
> > > So here I just try to follow the current code, see how lw is changed to c_lw.
> > 
> > 
> > Not exactly related to this PR, but I also saw a strange behaviour from MacroAssembler's lwu. it was generating lw + and ( a kind of lwu emulation) instead of lwu
> > an example
> > ```
> >   0.44%  ?  0x0000003fa46a86c8:   slli    t3,t3,0x20
> >    0.48%  ?  0x0000003fa46a86ca:   addi    t3,t3,-1
> >   ....
> >    3.11%  ?  0x0000003fa46a86dc:   lw    a0,0(t1)
> >    5.34%  ?  0x0000003fa46a86e0:   and    a0,a0,t3
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Using Assembler::lwu directly resulted in a correctly generated lwu
> 
> Interesting. This does not seem to reflect on the code of `MacroAssembler's lwu`. I wonder how could that happen.

If you take this PR https://github.com/openjdk/jdk/pull/17046/files#diff-7a5c3ed05b6f3f06ed1c59f5fc2a14ec566a6a5bd1d09606115767daa99115bdR3717 and change explicit Assembler::lwu() to lwu() then you are likely to see this issue

**Comment by zifeihan** [comment]
> > > > > We already have "macroses" for load and stores in macroAssembler_riscv.hpp, what's the reason to do compression decision in assembler_riscv.hpp instead ( not saying it's wrong) ?
> > > > > https://github.com/openjdk/jdk/blob/38d94725a1a85156e30b72b325886b0e25d4db03/src/hotspot/cpu/riscv/macroAssembler_riscv.hpp#L880
> > > > 
> > > > 
> > > > No, you are correct I also think this is not optimal. I don't know the background, but it seems like this is the easiest way to add compressed transparently. But to fully utilize C instruction we should favor the x8->x15, we often don't get C due to e.g. BCP is in x22. I think to be able to better utilize C we can't have it so transparent.
> > > > So here I just try to follow the current code, see how lw is changed to c_lw.
> > > 
> > > 
> > > Not exactly related to this PR, but I also saw a strange behaviour from MacroAssembler's lwu. it was generating lw + and ( a kind of lwu emulation) instead of lwu
> > > an example
> > > ```
> > >   0.44%  ?  0x0000003fa46a86c8:   slli    t3,t3,0x20
> > >    0.48%  ?  0x0000003fa46a86ca:   addi    t3,t3,-1
> > >   ....
> > >    3.11%  ?  0x0000003fa46a86dc:   lw    a0,0(t1)
> > >    5.34%  ?  0x0000003fa46a86e0:   and    a0,a0,t3
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Using Assembler::lwu directly resulted in a correctly generated lwu
> > 
> > 
> > Interesting. This does not seem to reflect on the code of `MacroAssembler's lwu`. I wonder how could that happen.
> 
> If you take this PR https://github.com/openjdk/jdk/pull/17046/files#diff-7a5c3ed05b6f3f06ed1c59f5fc2a14ec566a6a5bd1d09606115767daa99115bdR3717 and change explicit Assembler::lwu() to lwu() then you are likely to see this issue

Hi, I have tried to use MacroAssembler::lwu instead, and I see low difference in stub code emitted.
I have added comment [1]. let's discuss on that PR.
[1] https://github.com/openjdk/jdk/pull/17046/files#r1432182715

**Comment by robehn** [comment]
Passes t1+t2 fastdebug (with some expected timeouts).

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 1]
```diff
@@ -537,9 +537,9 @@ class Assembler : public AbstractAssembler {
   }
 
   INSN(lb,  0b0000011, 0b000);
-  INSN(lbu, 0b0000011, 0b100);
-  INSN(lh,  0b0000011, 0b001);
-  INSN(lhu, 0b0000011, 0b101);
+  INSN(_lbu, 0b0000011, 0b100); // Zcb
+  INSN(_lh,  0b0000011, 0b001); // Zcb
+  INSN(_lhu, 0b0000011, 0b101); // Zcb
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:542]: The code comment for these three lines seems misleading. These are normal 4-bytes encoding load/store instructions, not `Zcb` compressed instructions.

File: src/hotspot/cpu/riscv/assembler_riscv.hpp [hunk 2]
```diff
@@ -2917,6 +2926,239 @@ enum Nf {
...
+    } else {
+      if (Unsigned) {
+        _lhu(Rd_Rs2, Rs1, uimm);
+      } else {
+        _lh(Rd_Rs2, Rs1, uimm);
+      }
+    }
+  }
+
+  // Format CU, c.[sz]ext.*, c.no
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/assembler_riscv.hpp:2962]: Nit: s/c.no/c.not/

**Review by RealFYang — COMMENTED** [review]
Seems fine. I only have some minor comments.

File: src/hotspot/cpu/riscv/assembler_riscv.hpp (hunk 1 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp:542]: The comment was suppose to mean, these are 'overridden' by zcb, not C as the others.
I'll clarify that.

File: src/hotspot/cpu/riscv/macroAssembler_riscv.hpp [hunk 3]
```diff
@@ -489,7 +493,12 @@ class MacroAssembler: public Assembler {
   }
 
   inline void zext_b(Register Rd, Register Rs) {
-    andi(Rd, Rs, 0xFF);
+    if (do_compress_zcb(Rd, Rs) &&
+        (Rd == Rs)) {
```
[the remark this answers is not in the document]
Comment (robehn) [reply src/hotspot/cpu/riscv/macroAssembler_riscv.hpp]: Fixed

File: src/hotspot/cpu/riscv/assembler_riscv.hpp (hunk 1 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp]: I remove them, fixed.

File: src/hotspot/cpu/riscv/assembler_riscv.hpp (hunk 2 above)
Comment (robehn) [reply src/hotspot/cpu/riscv/assembler_riscv.hpp]: Fixed

**Comment by robehn** [comment]
> Seems fine. I only have some minor comments.

Thank you!

**Review by VladimirKempik — APPROVED** [review]

**Review by RealFYang — APPROVED** [review]
Updated change looks good. Thanks.

**Comment by robehn** [comment]
Thank you @VladimirKempik @RealFYang

**Pull request closed without merging**